File: week-07/rimrim990/광고 삽입.py
def solution(play_time, adv_time, logs):
  play_time = to_ss(play_time)
  adv_time = to_ss(adv_time)

  cnt = [0 for _ in range(play_time+1)]
  for log in logs:
    start, end = log.split("-")
    start, end = to_ss(start), to_ss(end)
    cnt[start] += 1
    cnt[end] -=1

  # 시청자 수
  for i in range(1, play_time+1):
    cnt[i] += cnt[i-1]

  # 시청자 수의 누적 합
  for i in range(1, play_time+1):
    cnt[i] += cnt[i-1]

  # 부분합이 최대가 되는 구간 탐색
  max_val, start = cnt[adv_time-1], 0
  for i in range(1, play_time-adv_time+1):
    val = cnt[i+adv_time-1] - cnt[i-1]
    if val > max_val:
      max_val = val
      start = i

  return to_str(start)

def to_str(time):
  ss = str(time % 60); time //= 60
  mm = str(time % 60); time //= 60
  hh = str(time)

  ss = ('0' + ss)[-2:]
  mm = ('0' + mm)[-2:]
  hh = ('0' + hh)[-2:]

  return ":".join([hh, mm, ss])

def to_ss(time):
  hh, mm, ss = time.split(":")
  return int(hh)*60*60 + int(mm)*60 + int(ss)

File: week-07/protoseo/광고 삽입.java
class Solution {

    long[] timeToViewer = new long[360001];

    public String solution(String play_time, String adv_time, String[] logs) {
        int advTime = convertTimeToInt(adv_time);

        for (String log : logs) {
            String[] split = log.split("-");
            int startTime = convertTimeToInt(split[0]);
            int endTime = convertTimeToInt(split[1]);

            timeToViewer[startTime]++;
            timeToViewer[endTime]--;
        }
        for (int i = 1; i <= 360000; i++) {
            timeToViewer[i] += timeToViewer[i - 1];
        }
        for (int i = 1; i <= 360000; i++) {
            timeToViewer[i] += timeToViewer[i - 1];
        }

        int answer = 0;
        long maxView = timeToViewer[advTime];
        for (int i = 1; i <= 360000 - advTime; i++) {
            long viewer = timeToViewer[i + advTime] - timeToViewer[i];
            if (maxView < viewer) {
                answer = i + 1;
                maxView = viewer;
            }
        }

        return convertIntToTime(answer);
    }

    String convertIntToTime(int value) {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("%02d", (value / 3600))).append(":");
        value %= 3600;
        sb.append(String.format("%02d", (value / 60))).append(":");
        value %= 60;
        sb.append(String.format("%02d", value));
        return sb.toString();
    }

    int convertTimeToInt(String time) {
        String[] result = time.split(":");

        return Integer.parseInt(result[0]) * 3600 + Integer.parseInt(result[1]) * 60 + Integer.parseInt(result[2]);
    }
}
